### PR TITLE
MINIFICPP-1755 - Use std::span instead of gsl::span

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/utils/gsl.h [hunk 1]
```diff
@@ -32,18 +33,23 @@ using remove_cvref_t = typename std::remove_cv<typename std::remove_reference<T>
...
+Container<detail::remove_cvref_t<T>> span_to(std::span<T> span) {
   static_assert(std::is_constructible<Container<detail::remove_cvref_t<T>>, typename gsl::span<T>::iterator, typename gsl::span<T>::iterator>::value,
       "The destination container must have an iterator (pointer) range constructor");
   return span_to<Container<detail::remove_cvref_t<T>>>(span);
 }
 
+template<typename T, typename U>
+std::span<T> as_span(std::span<U> value) {
+  return std::span{reinterpret_cast<T*>(value.data()), value.size_bytes() / sizeof(T)};
+}
```
Comment (fgerlits) [inline libminifi/include/utils/gsl.h:51]: the `span_to` and `as_span` functions could be moved to a new file, eg. `utils/span.h`, as they don't belong in `gsl.h` any longer

File: libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp [hunk 2]
```diff
@@ -67,7 +67,7 @@ void LineByLineInputOutputStreamCallback::readLine() {
   if (end_of_line != input_.end()) { ++end_of_line; }
 
   current_line_ = next_line_;
-  next_line_ = utils::span_to<std::string>(gsl::make_span(&*current_pos_, &*end_of_line).as_span<char>());
+  next_line_ = utils::span_to<std::string>(utils::as_span<char>(std::span(&*current_pos_, &*end_of_line)));
```
Comment (fgerlits) [inline libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp:70]: I think we could remove the `&*`s:
```suggestion
  next_line_ = utils::span_to<std::string>(utils::as_span<char>(std::span(current_pos_, end_of_line)));
```
as there is a `span` constructor which takes a pair of iterators

File: libminifi/include/utils/crypto/ciphers/Aes256Ecb.h [hunk 3]
```diff
@@ -50,8 +50,8 @@ class Aes256EcbCipher {
   static constexpr size_t KEY_SIZE = 32;
 
   explicit Aes256EcbCipher(Bytes encryption_key);
-  void encrypt(gsl::span<unsigned char /*, BLOCK_SIZE*/> data) const;
-  void decrypt(gsl::span<unsigned char /*, BLOCK_SIZE*/> data) const;
+  void encrypt(std::span<unsigned char /*, BLOCK_SIZE*/> data) const;
+  void decrypt(std::span<unsigned char /*, BLOCK_SIZE*/> data) const;
```
Comment (szaszm) [inline libminifi/include/utils/crypto/ciphers/Aes256Ecb.h:54]: Maybe we could change to a fixed extent span here?

File: libminifi/test/TestBase.cpp [hunk 4]
```diff
@@ -645,7 +645,7 @@ std::string TestPlan::getContent(const minifi::core::FlowFile& file) const {
   auto content_stream = content_repo_->read(*content_claim);
   auto output_stream = std::make_shared<minifi::io::BufferStream>();
   minifi::InputStreamPipe{*output_stream}(content_stream);
-  return utils::span_to<std::string>(output_stream->getBuffer().as_span<const char>());
+  return std::string{reinterpret_cast<const char*>(output_stream->getBuffer().data()), output_stream->size()};
```
Comment (szaszm) [inline libminifi/test/TestBase.cpp:648]: I'd prefer to keep following the as_span<const char> -> span_to<string> pattern here as well.

File: libminifi/test/rocksdb-tests/SwapTests.cpp [hunk 5]
```diff
@@ -57,7 +57,7 @@ class OutputProcessor : public core::Processor {
     auto ff = session->create();
     ff->addAttribute("index", id);
     session->write(ff, [&] (const std::shared_ptr<minifi::io::OutputStream>& output) -> int64_t {
-      auto ret = output->write(gsl::span<const char>(id.data(), id.size()).as_span<const std::byte>());
+      auto ret = output->write(as_bytes(std::span<const char>(id.data(), id.size())));
```
Comment (szaszm) [inline libminifi/test/rocksdb-tests/SwapTests.cpp:60]: Could we skip the explicit `std::span` construction, and just rely on implicit conversion from `string` to `span` in the argument of `as_bytes`?
```suggestion
      auto ret = output->write(as_bytes(id));
```

File: libminifi/test/unit/GslTest.cpp [hunk 6]
```diff
@@ -1,35 +0,0 @@
...
-
-TEST_CASE("span to", "[span to]") {
-  const auto test_span = gsl::make_span("test text", 9);
-  const auto string = utils::span_to<std::string>(test_span);
-  const auto vector = utils::span_to<std::vector>(test_span);
-
-  REQUIRE(string == "test text");
-  REQUIRE('t' == vector[0]);
-  REQUIRE(9 == vector.size());
-}
```
Comment (szaszm) [inline libminifi/test/unit/GslTest.cpp:35]: why is this deleted? we should also add tests for as_span.

**Review by fgerlits — APPROVED** [review]

File: libminifi/test/TestBase.cpp (hunk 4 above)
Comment (adamdebreceni) [reply libminifi/test/TestBase.cpp]: done

File: libminifi/include/utils/gsl.h [hunk 7]
```diff
@@ -17,6 +17,7 @@
 
 #pragma once
 
+#include <span>
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/utils/gsl.h:20]: removed

File: libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp (hunk 2 above)
Comment (adamdebreceni) [reply libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp]: I had to revert this as there doesn't seems to be an iterator overload constructor in the macos ci runner

**Review by martinzink — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]
LGTM, just a really minor suggestion

File: libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp [hunk 8]
```diff
@@ -67,7 +68,7 @@ void LineByLineInputOutputStreamCallback::readLine() {
   if (end_of_line != input_.end()) { ++end_of_line; }
 
   current_line_ = next_line_;
-  next_line_ = utils::span_to<std::string>(gsl::make_span(&*current_pos_, &*end_of_line).as_span<char>());
+  next_line_ = utils::span_to<std::string>(utils::as_span<char>(std::span(std::to_address(current_pos_), std::to_address(end_of_line))));
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp:71]: according to the requirements of [contiguous_iterator](https://en.cppreference.com/w/cpp/iterator/contiguous_iterator) it should work

**Review by szaszm — APPROVED** [review]
